File: backend/app/platform/config/knowledge_scopes.py
"""Loads the recognized knowledge-scope tags (PLAN 10, ADR-0011) from a dedicated config file.

Deliberately a dedicated file, not an env var: this is a content registry (which integration
platforms exist) an operator edits directly, not a per-deployment secret or runtime toggle —
those stay in `.env` (`enable_knowledge_scope_filtering`, `default_knowledge_scope`). It lives at
`knowledge-base/config/` so it reads as one global place regardless of which app you're looking at
— `frontend` reads the same file (the PLAN 10.8 scope-switcher UI generates its list from this JSON
at build time, substep 1.2.2) instead of duplicating the list.

Validation (substep 1.2.1, panels tg-config / ks-validate / cm-config). The file decides what a
person may see, so a bad entry stops the backend before it can leak. At startup, in order, each
rule raises with a message that names the offending entry:
  1. every entry is lowercase (an uppercase slug is rejected, not silently lowercased);
  2. no duplicates;
  3. the general scope (`obi-general-test`) is present;
  4. the reserved `classified` scope is present.
`classified` is reserved: it must be in the file, but it is never a recognized/offerable scope —
it is excluded from the returned set, from GET /health, and from the widget's scope list. A body
`knowledge_scope: "classified"` therefore stays an unknown-scope 400, never selectable.
"""

from __future__ import annotations

import json
from pathlib import Path

# backend/app/platform/config/knowledge_scopes.py -> repo root is 4 levels up.
DEFAULT_KNOWLEDGE_SCOPES_PATH = (
    Path(__file__).resolve().parents[4] / "knowledge-base" / "config" / "knowledge_scopes.json"
)

# The always-present base scope (ADR-0011): untagged content falls into it and it is always allowed.
GENERAL_SCOPE = "obi-general-test"
# Reserved scope: present in the file, never recognized/offered. Kept in sync with the forbidden
# integration slug in platform/config/platforms.py (`_FORBIDDEN`).
RESERVED_CLASSIFIED_SCOPE = "classified"


def load_recognized_knowledge_scopes(path: Path = DEFAULT_KNOWLEDGE_SCOPES_PATH) -> frozenset[str]:
    """Validate the tag file at startup and return the recognized, offerable scopes.

    The returned set is every scope in the file EXCEPT the reserved `classified` scope. A failing
    rule raises ``ValueError`` before ``create_app`` finishes, so a malformed file exits the
    process non-zero rather than serving one request under wrong permissions.
    """
    data = json.loads(path.read_text())
    names: list[str] = []
    for entry in data["scopes"]:
        name = str(entry["name"]).strip()
        if name:
            names.append(name)

    # 1. every entry lowercase — name the offending entry.
    for name in names:
        if name != name.lower():
            raise ValueError(f"{path}: knowledge scope {name!r} must be lowercase")

    # 2. no duplicates — name the offending entry.
    seen: set[str] = set()
    for name in names:
        if name in seen:
            raise ValueError(f"{path}: knowledge scope {name!r} is listed more than once")
        seen.add(name)

    # 3. the general scope present.
    if GENERAL_SCOPE not in seen:
        raise ValueError(f"{path} must include the {GENERAL_SCOPE!r} scope (got {sorted(seen)!r})")

    # 4. the reserved classified scope present (never offered — excluded from the returned set).
    if RESERVED_CLASSIFIED_SCOPE not in seen:
        raise ValueError(
            f"{path} must include the reserved {RESERVED_CLASSIFIED_SCOPE!r} scope "
            f"(got {sorted(seen)!r})"
        )

    return frozenset(seen - {RESERVED_CLASSIFIED_SCOPE})

File: frontend/src/features/chat/model/knowledge-scopes.ts
/**
 * knowledge-scopes — the recognized RAG knowledge scopes the dev/verification scope switcher
 * (PLAN 10.8) lets a tester pick between.
 *
 * GENERATED, NOT hand-written (substep 1.2.2, panels ks-deploy / ks-widget / cm-config). This list
 * is built at import/build time from the ONE canonical allowlist,
 * `knowledge-base/config/knowledge_scopes.json` — the same file the backend reads at startup
 * (ADR-0011, PLAN 10.1) and Confluence ingestion reads for scope eligibility. There is no second
 * source of truth and no hand-maintained copy: to recognize a new scope, add one entry (with its
 * `label`) to that JSON and rebuild — nothing here changes. The reserved `classified` scope is
 * present in the JSON but filtered out below, so it is never offered by the widget and can never
 * ride on a request body (ADR-0011).
 *
 * The JSON is reached through the `@kb/*` tsconfig path alias (mirrored in `vitest.config.ts`), so
 * the relative depth to the repo-root config is written once, in the alias, not here.
 */
import scopeConfig from "@kb/config/knowledge_scopes.json";

// Reserved scope: present in the canonical file so the backend knows it, but never offered by the
// widget. Kept in sync with the backend loader's RESERVED_CLASSIFIED_SCOPE and the forbidden
// integration slug (ADR-0011).
const RESERVED_CLASSIFIED_SCOPE = "classified";

/**
 * A recognized, offerable scope's tag name. Widened to `string` now that the set is generated from
 * the JSON at build time (the runtime backend gate — not a compile-time union — is what rejects an
 * unknown scope). The exported name is unchanged from the pre-1.2.2 hand-written list, so no
 * importer changes.
 */
export type KnowledgeScope = string;

/**
 * The offerable scopes, in the JSON's order, each as `{ name, label }` — `classified` excluded.
 * `name` is the exact Confluence tag / internal scope id; `label` is the human-readable name the
 * switcher displays. Both come straight from the canonical JSON.
 */
export const KNOWLEDGE_SCOPES: readonly { name: KnowledgeScope; label: string }[] =
  scopeConfig.scopes
    .filter((scope) => scope.name !== RESERVED_CLASSIFIED_SCOPE)
    .map((scope) => ({ name: scope.name, label: scope.label }));

export const KNOWLEDGE_SCOPE_NAMES: readonly KnowledgeScope[] = KNOWLEDGE_SCOPES.map((s) => s.name);
